File: src/api/adminEventRepository.ts
import { supabase } from '../../database/supa/supabase'

export interface SharedAdminEvent {
  id: string
  source: 'admin'
  courseCode: string
  subject: string
  date: string
  startMinutes: number
  endMinutes: number
  classType: string
  section: string
  room: string
  studentCount: string
  instructorLastName: string
}

interface AdminEventRow {
  id: string
  title: string
  event_date: string
  room: string
  start_minutes: number
  end_minutes: number
}

export async function loadSharedAdminEvents(): Promise<SharedAdminEvent[]> {
  if (!supabase) return []

  const { data, error } = await supabase
    .from('admin_events')
    .select('id,title,event_date,room,start_minutes,end_minutes')
    .order('updated_at')

  if (error) throw error

  return (data as AdminEventRow[]).map(row => ({
    id: row.id,
    source: 'admin',
    courseCode: row.title,
    subject: '',
    date: row.event_date,
    startMinutes: row.start_minutes,
    endMinutes: row.end_minutes,
    classType: 'EVENT',
    section: '',
    room: row.room,
    studentCount: '',
    instructorLastName: '',
  }))
}

export async function saveSharedAdminEvent(event: {
  id: string
  courseCode: string
  date?: string
  room: string
  startMinutes: number
  endMinutes: number
}): Promise<void> {
  if (!supabase) return
  if (!event.date) throw new Error('An event date is required.')

  const { error } = await supabase
    .from('admin_events')
    .upsert({
      id: event.id,
      title: event.courseCode,
      event_date: event.date,
      room: event.room,
      start_minutes: event.startMinutes,
      end_minutes: event.endMinutes,
      updated_at: new Date().toISOString(),
    }, { onConflict: 'id' })

  if (error) throw error
}

export async function deleteSharedAdminEvent(id: string): Promise<void> {
  if (!supabase) return

  const { error } = await supabase
    .from('admin_events')
    .delete()
    .eq('id', id)

  if (error) throw error
}

export function subscribeToSharedAdminEvents(onChange: () => void) {
  const client = supabase
  if (!client) return () => undefined

  const channel = client
    .channel('admin-event-changes')
    .on(
      'postgres_changes',
      {
        event: '*',
        schema: 'public',
        table: 'admin_events',
      },
      onChange,
    )
    .subscribe()

  return () => {
    void client.removeChannel(channel)
  }
}

File: backend/database/supabase/admin_event_repository.py
from typing import Any

from .client import request

PAGE_SIZE = 1000


def _load_event_rows(select: str) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    offset = 0
    while True:
        page = request("GET", "admin_events", query={
            "select": select,
            "order": "updated_at.asc",
            "limit": str(PAGE_SIZE),
            "offset": str(offset),
        }) or []
        rows.extend(page)
        if len(page) < PAGE_SIZE:
            return rows
        offset += PAGE_SIZE


def load_events() -> list[dict[str, Any]]:
    try:
        rows = _load_event_rows(
            "id,title,event_date,room,start_minutes,end_minutes,assistant_id,assistant_label,updated_at",
        )
    except RuntimeError:
        rows = _load_event_rows(
            "id,title,event_date,room,start_minutes,end_minutes,updated_at",
        )
    return [{
        "id": row["id"],
        "source": "admin",
        "courseCode": row["title"],
        "subject": "",
        "date": row["event_date"],
        "startMinutes": row["start_minutes"],
        "endMinutes": row["end_minutes"],
        "classType": "EVENT",
        "section": "",
        "room": row["room"],
        "studentCount": "",
        "instructorLastName": "",
        "assistantId": row.get("assistant_id") or None,
        "assistantLabel": row.get("assistant_label") or None,
    } for row in rows]


def save_event(event: dict[str, Any]) -> None:
    payload = {
        "id": event["id"],
        "title": event["courseCode"],
        "event_date": event["date"],
        "room": event["room"],
        "start_minutes": event["startMinutes"],
        "end_minutes": event["endMinutes"],
        "assistant_id": event.get("assistantId") or None,
        "assistant_label": event.get("assistantLabel") or None,
    }
    try:
        request("POST", "admin_events", payload=payload, prefer="resolution=merge-duplicates,return=minimal")
    except RuntimeError:
        payload.pop("assistant_id", None)
        payload.pop("assistant_label", None)
        request("POST", "admin_events", payload=payload, prefer="resolution=merge-duplicates,return=minimal")


def delete_event(event_id: str) -> None:
    request("DELETE", "admin_events", query={"id": f"eq.{event_id}"})
